cpp/include/cugraph/algorithms.hpp

```diff
@@ -1893,6 +1893,14 @@ k_core(raft::handle_t const& handle,
  * level
  * @param rng_state A pre-initialized raft::RngState object for generating random numbers
  * @param return_hops boolean flag specifying if the hop information should be returned
+ * @param unique_sources boolean flag, if true then sources at each hop will not be repeated.  That
+ * is, if vertex v appears as a source in hop X, it will never be sent as a source to a hop > X.
+ * Default is false.
+ * @param carry_over_sources boolean flag, if true then sources for hop X will be repeated as
+ * sources for all hops > X.  Default is false
```

Comment:
Maybe we should just throw an exception.  Those flags are incompatible, after all.

Reply:
In that case, we may make it impossible to make both true.

Something like
```
enum class some_name {
  unique_sources = 0,
  carry_over_sources = 1,
  neither (we need a better name for this) = 2
};
```

Reply:
So, conceptually, we have three options.

1. Carry over sources,
2. Do not carry over sources,
3. Do not carry over sources and also do not allow previously appeared sources to appear again.

After this, we have an option to perform dedupe or not.

The code can be

if 1, append the sources
else if 2, do nothing
else (i.e. 3), remove sources that have already appeared.

Then if (dedupe) perform deduplication.

Reply:
I think `unique_sources` here is a misnomer (why do we have `unique_sources` and `dedupe_sources`? We need to dig into the code to figure out the answer). `unique_sources` here really means do not use any sources that have appeared in previous hops.

Reply:
well, hang on, carry over and deduplicate can happen together

Reply:
What should happen in that case is if vertex 3 appeared in the previous destinations, and it also appeared in sources earlier, it should appear twice.

Reply:
And I did verify that behavior when I was testing

Reply:
So the only change to your logic would be where the dedupe happens.  Otherwise, yes, that's correct.

Reply:
@VibhuJawa can you confirm that I have the expected behavior right?

Reply:
>well, hang on, carry over and no-deduplicate can happen together

Yes, this happens when one picks carry_over and set dedupe to false.

> What should happen in that case is if vertex 3 appeared in the previous destinations, and it also appeared in sources earlier, it should appear twice.
> So the only change to your logic would be where the dedupe happens. Otherwise, yes, that's correct.

You mean this should be

```
if (dedupe) de-duplicate next iteration sources

if 1, append the sources
else if 2, do nothing
else (i.e. 3), remove sources that have already appeared.
```
?

Reply:
Yep, that's it.  But we should confirm with Vibhu.

Reply:
>What should happen in that case is if vertex 3 appeared in the previous destinations, and it also appeared in sources earlier, it should appear twice.

I am confused by this. 

>What happens if unique_sources is true and carry_over_sources is true? Better document the expected behavior in this case?

Lets say we have : 

```
1 2
3 2
2 1
```
 
 We get `2, 2, 1` as new sources.
 
Then we carry over sources to it so we get ` 2, 2, 1, 3,1,2` which we deuplicate to get `1, 2,3`. 

>Maybe we should just throw an exception. Those flags are incompatible, after all.

Why is this unexpected.

Reply:
Running the above example through dgl to verify

Reply:
It's unexpected since `unique_sources` is undoing `carry_over_sources`

Reply:
This is the behavior we want. 

```python3
g = dgl.graph(([1, 2, 3], [2, 2, 1]))
seed_nodes = [1,2]

for hop_id,fanout in enumerate([1,1]):
    print(f"hop_id: {hop_id} seed nodes {seed_nodes}")
    sg = g.sample_neighbors(seed_nodes, fanout)
    print("sampled graph", sg.edges())
    # Convert this subgraph to a message flow graph.
    sg = dgl.to_block(sg, seed_nodes)
    seed_nodes = sg.srcdata[dgl.NID]
    print("--"*30)
 ```
 
 ```python3
hop_id: 0 seed nodes [1, 2]
sampled graph (tensor([3, 2]), tensor([1, 2]))
------------------------------------------------------------
hop_id: 1 seed nodes tensor([1, 2, 3])
sampled graph (tensor([3, 2]), tensor([1, 2]))
------------------------------------------------------------
```

Reply:
ok, so the dedupe goes at the end the, the way @seunghwak originally had it.

Reply:
>It's unexpected since unique_sources is undoing carry_over_sources. Sources that are carried over will always be removed if unique_sources is True.

Why will all the **sources that are carried** over be removed ?

Reply:
because if they were carried over, they would have been sources in a previous hop

Reply:
>because if they were carried over, they would have been sources in a previous hop

I guess i am confused by what unique sources vs dedup sources refers too here. 
> The code can be
> 
> if 1, append the sources
> else if 2, do nothing
> else (i.e. 3), remove sources that have already appeared.
> 
> Then if (dedupe) perform deduplication.

This seems fine to me. .

Reply:
`unique_sources`: A source can never be a source in a future hop
`dedupe_sources`: Deduplicates the source list before performing sampling

Reply:
I'm thinking if C++ changes to use an enum, we'll use something similar at the Python level (i.e.)

sources='default'
sources='carryover'
sources='unique'

Reply:
So...
* Add an enum `cugraph_sampling_source_behavior_t` with values: `default`, `carryover`, `unique` (and a corresponding C++ enum)
* Leave the `dedupe_sources` flag as is
* Remove the boolean `unique_sources` and `carry_over_sources` and pass a `cugraph_sampling_source_behavior_t` to the C API flag, as a parameter in the C++ API
* Modify the boolean switches for `unique_sources` and `carry_over_sources` to use the enum for control instead.

Would that resolve these issues satisfactorily?

Reply:
I still think it's not duplicating in the right place.  Based on what Vibhu said, it should duplicate right before sampling, after all the other flags have been applied.

Reply:
I think `unique` here is a misnomer. Unique can be interpreted as "no duplicates". Unique here really means do not use the sources that have already used in previous hops.

Reply:
how about `once`?

Reply:
How about `unique_across_hops`?

Reply:
This is the opposite of "carry_over_sources", may be "carry_over_prior_sources" and "exclude_prior_sources"?

Reply:
source_behavior='default'
source_behavior='carryover'
source_behavior='exclude_prior'

Reply:
If we decide to use "prior_sources", we need to rename

```
  raft::device_span<vertex_t const> starting_vertices,
  std::optional<raft::device_span<label_t const>> starting_vertex_labels,
 ```
 
accordingly as well. Starting vertices here are prior vertices, right? Or we may rename these variables to

"source_vertices" and "source_vertex_labels" and use add `enum class prior_sources_t` (we already have `cugraph_cc_t` and `k_core_degree_type_t`) with `default`, `carryover`, and `exclude`?

Reply:
the start vertices are priors, yeah

Reply:
OK, so `prior_sources` in the enum name, `exclude` instead of `unique`.  Sound reasonable for me to make the changes?

Reply:
(I'll try to ripple everything through consistently.)

Reply:
Fixed in next push